### ISSUE TITLE
Add specialized chunks for every primitive type

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#3387]
In order to facilitate box-less operations, we can add chunks for different primitive types. Technically, we should support not just `Array` but also NIO buffers (byte buffer, etc). 

For organizational purposes, these should probably be placed inside an object inside the companion object of Chunk, e.g.:

```scala
object Chunk {
  object Byte {
    final class ByteChunkArr(array: Array[Byte]) extends Chunk[A] {
      override def byte(index: Long): Byte = array(index)
    }
    final class ByteChunkBuf(buffer: ByteBuffer) extends Chunk[A] {
      ..
    }
  }
}
```

Specialized chunk subtypes will have efficient and box-free implementations of primitive-specialized methods.

See #3386

**Comment by vasilmkd** [comment]
With `Chunk` now being an `IndexedSeq` what do we stand to gain from this change (which was turned down ~6 months ago due to too much code duplication and a burden on maintainability)?

**Comment by jdegoes** [comment]
Any specialized methods we add (which are not inside `IndexedSeq`) do not need to box.

**Comment by michalwarecki** [comment]
I can take it

**Comment by iravid** [comment]
Resolved in #3576.
